signer/src/stacks/wallet.rs

```diff
@@ -0,0 +1,537 @@
+use blockstack_lib::chainstate::stacks::OrderIndependentMultisigSpendingCondition;
+use blockstack_lib::chainstate::stacks::StacksTransaction;
+use blockstack_lib::chainstate::stacks::TransactionAnchorMode;
+use blockstack_lib::chainstate::stacks::TransactionAuth;
+use blockstack_lib::chainstate::stacks::TransactionAuthFlags;
+use blockstack_lib::chainstate::stacks::TransactionPayload;
+use blockstack_lib::chainstate::stacks::TransactionPublicKeyEncoding;
+use blockstack_lib::chainstate::stacks::TransactionSpendingCondition;
+use blockstack_lib::chainstate::stacks::TransactionVersion;
+use blockstack_lib::core::CHAIN_ID_MAINNET;
+use blockstack_lib::core::CHAIN_ID_TESTNET;
+use blockstack_lib::types::chainstate::StacksAddress;
+use blockstack_lib::util::secp256k1::MessageSignature;
+use blockstack_lib::util::secp256k1::Secp256k1PublicKey;
+use secp256k1::ecdsa::RecoverableSignature;
+use secp256k1::Message;
+use secp256k1::PublicKey;
+use secp256k1::SecretKey;
+use secp256k1::SECP256K1;
+
+use crate::config::NetworkKind;
+use crate::error::Error;
+use crate::stacks::contracts::AsContractCall;
+
+/// Requisite info for the signers' multi-sig wallet on Stacks.
+#[derive(Debug, Clone)]
+pub struct SignerWallet {
+    /// The current set of public keys for all known signers during this
+    /// PoX cycle. These values must be sorted.
+    public_keys: Vec<PublicKey>,
+    /// The number of signers necessary for successfully signing a
+    /// multi-sig transaction.
+    signatures_required: u16,
+    /// The kind of network we are operating under.
+    network_kind: NetworkKind,
+}
+
+impl SignerWallet {
+    /// Create the wallet for the signer.
+    ///
+    /// # Note
+    ///
+    /// An error is returned if the
+    /// number of required signatures is greater than the list of public
+    /// keys or if the list of public keys is empty.
+    pub fn new(
+        mut public_keys: Vec<PublicKey>,
+        signatures_required: u16,
+        network_kind: NetworkKind,
+    ) -> Result<Self, Error> {
+        public_keys.sort();
+
+        if public_keys.len() < signatures_required as usize || public_keys.is_empty() {
+            return Err(Error::MissingBlock);
+        }
+
+        if signatures_required == 0 || public_keys.is_empty() {
+            return Err(Error::MissingBlock);
+        }
+
+        Ok(Self {
+            public_keys,
+            signatures_required,
+            network_kind,
+        })
+    }
+
+    fn hash_mode() -> OrderIndependentMultisigHashMode {
+        OrderIndependentMultisigHashMode::P2WSH
+    }
+
+    /// Return the stacks address for the signers
+    pub fn address(&self) -> StacksAddress {
+        let public_keys: Vec<Secp256k1PublicKey> = self
+            .public_keys
+            .iter()
+            .map(|pk| Secp256k1PublicKey::from_slice(&pk.serialize()))
+            .collect::<Result<_, _>>()
+            .expect("we know these are all valid public keys");
```

Comment:
It seems to me like this conversion happens whenever we use `public_keys` within this context. Because the `SignerWallet` is a stacks-specific type, perhaps we should consider converting the public keys once in `SignerWallet::new` and internally store the Stacks variant of the `Secp256k1PublicKey` type instead.

Reply:
Hmmm, good point. I'll give it a look. I'll change it if it ends up making things simpler.

Also, my motivation for using these types was that I wanted to rely on only one Secp256k1 library for internal use and do the conversions when necessary. But that decision might have lead to some incidental complexity so let me check.

Reply:
So it looks like it is the same complexity with both approaches. When storing the `Secp256k1PublicKey` you get:
1. More complex `SignerWallet::new`.
1. Simpler `StacksAddress` creation.
2. More complex `MultisigTx::push_signature`.
3. Simpler `MultisigTx::finalize_transaction`.

Seems like a wash, at least for now.